lib/scheduler.ml

```diff
@@ -0,0 +1,198 @@
+           if r <> [] then ignore(Unix.read pipe_out (String.create 1) 0 1);
+           (* return true if we waited the full length of time, false if we were woken *)
+           r = []
+         with Pre_signalled -> false
+      )
+      (fun () ->
+         mutex_execute x.m
+           (fun () ->
+              x.pipe_out <- None;
+              x.pipe_in <- None;
+              List.iter close' !to_close)
+      )
+
+  let signal (x: t) =
+    mutex_execute x.m
+      (fun () ->
+         match x.pipe_in with
+         | Some fd -> ignore(Unix.write fd "X" 0 1)
+         | None -> x.signalled <- true 	 (* If the wait hasn't happened yet then store up the signal *)
+      )
+end
+
+type item = {
+  id: int;
+  name: string;
+  fn: unit -> unit
+}
+
+let schedule = ref Int64Map.empty
+let delay = Delay.make ()
+let next_id = ref 0
+let m = Mutex.create ()
+
+type time =
+  | Absolute of int64
+  | Delta of int [@@deriving rpc]
+
+type t = int64 * int [@@deriving rpc]
+
+let now () = Unix.gettimeofday () |> ceil |> Int64.of_float
+
+module Dump = struct
+  type u = {
+    time: int64;
+    thing: string;
+  } [@@deriving rpc]
+  type t = u list [@@deriving rpc]
+  let make () =
+    let now = now () in
+    mutex_execute m
+      (fun () ->
+         Int64Map.fold (fun time xs acc -> List.map (fun i -> { time = Int64.sub time now; thing = i.name }) xs @ acc) !schedule []
+      )
+end
+
+let one_shot time (name: string) f =
+  let time = match time with
+    | Absolute x -> x
+    | Delta x -> Int64.(add (of_int x) (now ())) in
+  let id = mutex_execute m
+      (fun () ->
+         let existing =
+           if Int64Map.mem time !schedule
+           then Int64Map.find time !schedule
+           else [] in
+         let id = !next_id in
+         incr next_id;
+         let item = {
+           id = id;
+           name = name;
+           fn = f
+         } in
+         schedule := Int64Map.add time (item :: existing) !schedule;
+         Delay.signal delay;
+         id
+      ) in
+  (time, id)
+
+let cancel (time, id) =
+  mutex_execute m
+    (fun () ->
+       let existing =
+         if Int64Map.mem time !schedule
+         then Int64Map.find time !schedule
+         else [] in
+       schedule := Int64Map.add time (List.filter (fun i -> i.id <> id) existing) !schedule
+    )
+
+let process_expired () =
+  let t = now () in
+  let expired =
+    mutex_execute m
+      (fun () ->
+         let expired, unexpired = Int64Map.partition (fun t' _ -> t' <= t) !schedule in
+         schedule := unexpired;
+         Int64Map.fold (fun _ stuff acc -> acc @ stuff) expired [] |> List.rev) in
+  (* This might take a while *)
+  List.iter
+    (fun i ->
+       try
+         i.fn ()
+       with e ->
+         debug "Scheduler ignoring exception: %s\n%!" (Printexc.to_string e)
+    ) expired;
+  expired <> [] (* true if work was done *)
+
+let rec main_loop () =
+  while process_expired () do () done;
```

Comment:
No, process_expired returns false if there was no work to do, and we progress to the delay.